File: src/Cam.cpp
#include "Cam.h"
#include "godot_cpp/classes/accessibility_server.hpp"
#include "godot_cpp/classes/compositor.hpp"
#include "godot_cpp/classes/compositor_effect.hpp"
#include "godot_cpp/classes/engine.hpp"
#include "godot_cpp/classes/input.hpp"
#include "godot_cpp/classes/mesh_instance3d.hpp"
#include "godot_cpp/classes/ref.hpp"
#include "godot_cpp/classes/rendering_server.hpp"
#include "godot_cpp/classes/scene_tree.hpp"
#include "godot_cpp/classes/shader_material.hpp"
#include "godot_cpp/classes/texture2d.hpp"
#include "godot_cpp/classes/texture2drd.hpp"
#include "godot_cpp/classes/texture_rect.hpp"
#include "godot_cpp/classes/viewport.hpp"
#include "godot_cpp/core/memory.hpp"
#include "godot_cpp/variant/string_name.hpp"
#include "godot_cpp/variant/utility_functions.hpp"
#include <cstdio>
#include <godot_cpp/core/class_db.hpp>
#include <string>

using namespace godot;

void Cam::_bind_methods() {
  ClassDB::bind_method(D_METHOD("OnQuitPressed"), &Cam::OnQuitPressed);
  ClassDB::bind_method(D_METHOD("OnRdy"), &Cam::OnRdy);
}

Cam::Cam() { std::printf("bjr je suis la cam"); }

Cam::~Cam() {}

void Cam::_ready() {
  button = get_node<Button>("%Button");
  label = get_node<Label>("%Label");
  cam = get_node<Camera3D>("%Camera3D");
  meshInstance = get_node<MeshInstance3D>("%MeshInstance3D");
  texture = get_node<TextureRect>("%TextureRect");
  set_physics_process(false);

  const Ref<Compositor> comp = cam->get_compositor();
  for (Ref<CompositorEffect> ce : comp->get_compositor_effects()) {
    if (ce->is_class("PainterCpp")) {
      pe = ce;
      break;
    }
  }

  if (pe == nullptr)
    UtilityFunctions::push_error("painterCpp is null");
  pe->connect("OnRdy", Callable(this, "OnRdy"));
  RenderingServer::get_singleton()->call_on_render_thread(
      Callable(*pe, "InitCompute"));
  cam->set_cull_mask(1 << 20);
  meshInstance->set_layer_mask(meshInstance->get_layer_mask() | 1 << 20);
  button->connect("pressed", Callable(this, "OnQuitPressed"));
}

void Cam::_physics_process(double delta) {
  if (Input::get_singleton()->is_action_pressed("right")) {
    meshInstance->rotate_y(1.7f * delta);
  } else if (Input::get_singleton()->is_action_pressed("left")) {
    meshInstance->rotate_y(-1.7f * delta);
  }
  if (Input::get_singleton()->is_action_pressed("up")) {
    meshInstance->rotate_x(1.7f * delta);
  } else if (Input::get_singleton()->is_action_pressed("down")) {
    meshInstance->rotate_x(-1.7f * delta);
  }
  const Viewport *v = get_viewport();
  cam->look_at(project_position(v->get_mouse_position(), 5.0f));
  std::string s =
      std::to_string(Engine::get_singleton()->get_frames_per_second());
  label->set_text(s.c_str());
}

void Cam::OnRdy(RID r) {
  UtilityFunctions::print("roger");
  Ref<Texture2DRD> t = memnew(Texture2DRD);
  t->set_texture_rd_rid(r);

  Ref<ShaderMaterial> sm = meshInstance->get_material_overlay();
  sm->set_shader_parameter("paintedTexture", t);
  set_physics_process(true);
}

void Cam::OnQuitPressed() {
  propagate_notification(NOTIFICATION_WM_CLOSE_REQUEST);
  get_tree()->quit();
}

File: Cam.cs
using Godot;

public partial class Cam : Camera3D
{
	static readonly StringName ActionRight = "right";
	static readonly StringName ActionLeft = "left";
	static readonly StringName ActionUp = "up";
	static readonly StringName ActionDown = "down";
	static readonly StringName PaintedTextureParam = "paintedTexture";

	const float ROTATE_SPEED = 1.7f;
	const float LOOK_AT_DEPTH = 5.0f;
	const double LABEL_REFRESH_SECONDS = 0.25;

	MeshInstance3D _meshInstance;

	ShaderMaterial _overlayMaterial;

	Painter pe;

	TextureRect _texture;

	Camera3D _cam;

	Label label;

	Button _button;

	Viewport _viewport;

	SubViewport _subViewport;

	Rid _viewportRid;

	Rid _subViewportRid;

	Texture2Drd _paintedTexture;

	Vector2 _lastMousePos = new(float.NaN, float.NaN);

	double _fpsAccum;

	public override void _Ready()
	{
		_button = GetNode<Button>("%Button");
		label = GetNode<Label>("%Label");
		_cam = GetNode<Camera3D>("%Camera3D");
		_subViewport = _cam.GetParent<SubViewport>();
		_meshInstance = GetNode<MeshInstance3D>("%MeshInstance3D");
		_overlayMaterial = _meshInstance.MaterialOverlay as ShaderMaterial;
		_texture = GetNode<TextureRect>("%TextureRect");
		_viewport = GetViewport();
		_viewportRid = _viewport.GetViewportRid();
		RenderingServer.ViewportSetMeasureRenderTime(_viewportRid, true);
		SetPhysicsProcess(false);

		foreach (CompositorEffect ce in _cam.Compositor.CompositorEffects)
		{
			if (ce is Painter p)
			{
				pe = p;
				break;
			}
		}

		if (pe == null)
		{
			GD.PrintErr("[Cam] Painter compositor effect not found on _cam.Compositor");
			return;
		}
		GD.Print("[Cam] Painter effect found, requesting compute init");

		if (LeakSentry.Instance != null)
		{
			LeakSentry.Instance.LeakSuspected += OnLeakSuspected;
			GD.Print("[Cam] leak_sentry armed: shift+L overlay, shift+G report, shift+D dump");
		}

		pe.OnRdy += OnRdy;
		RenderingServer.CallOnRenderThread(Callable.From(pe.InitCompute));
		_cam.CullMask = 1 << 20;
		_meshInstance.Layers |= 1 << 20;

		_button.Pressed += OnQuitPressed;
	}

	public void OnRdy(Rid r)
	{
		_paintedTexture = new Texture2Drd { TextureRdRid = r };
		LeakSentry.Track(_paintedTexture, "painted-texture");
		_overlayMaterial?.SetShaderParameter(PaintedTextureParam, _paintedTexture);
		_texture.Texture = _paintedTexture;
		if (_subViewport != null)
		{
			_subViewportRid = _subViewport.GetViewportRid();
			_subViewport.RenderTargetUpdateMode = SubViewport.UpdateMode.Disabled;
			RenderingServer.ViewportSetMeasureRenderTime(_subViewportRid, true);
			GD.Print("[Cam] brush viewport gated on input (renders only on rotate/mouse-move)");
		}
		else
		{
			GD.PrintErr("[Cam] SubViewport not found, brush renders every frame (no gating)");
		}
		SetPhysicsProcess(true);
		GD.Print("[Cam] painted texture bound to mesh overlay, input enabled");
	}

	void OnLeakSuspected(string metric, double growthPerSecond)
	{
		GD.PrintErr($"[Cam] leak_sentry suspects {metric} growing at {growthPerSecond:0.00}/s");
	}

	public override void _ExitTree()
	{
		if (LeakSentry.Instance != null)
			LeakSentry.Instance.LeakSuspected -= OnLeakSuspected;

		if (pe == null)
			return;
		_overlayMaterial?.SetShaderParameter(PaintedTextureParam, default);
		_paintedTexture = null;
		RenderingServer.CallOnRenderThread(Callable.From(pe.Cleanup));
		GD.Print("[Cam] requested Painter GPU cleanup on render thread");
	}

	public override void _PhysicsProcess(double delta)
	{
		float step = ROTATE_SPEED * (float)delta;
		bool changed = false;

		if (Input.IsActionPressed(ActionRight))
		{
			_meshInstance.RotateY(step);
			changed = true;
		}
		else if (Input.IsActionPressed(ActionLeft))
		{
			_meshInstance.RotateY(-step);
			changed = true;
		}

		if (Input.IsActionPressed(ActionUp))
		{
			_meshInstance.RotateX(step);
			changed = true;
		}
		else if (Input.IsActionPressed(ActionDown))
		{
			_meshInstance.RotateX(-step);
			changed = true;
		}

		Vector2 mouse = _viewport.GetMousePosition();
		bool mouseMoved = mouse != _lastMousePos;

		if (mouseMoved)
		{
			_cam.LookAt(ProjectPosition(mouse, LOOK_AT_DEPTH));
			_lastMousePos = mouse;
		}

		if (_subViewport != null && (changed || mouseMoved))
			_subViewport.RenderTargetUpdateMode = SubViewport.UpdateMode.Once;

		_fpsAccum += delta;
		if (_fpsAccum < LABEL_REFRESH_SECONDS)
			return;
		_fpsAccum = 0.0;


		double gpuMs = RenderingServer.ViewportGetMeasuredRenderTimeGpu(_viewportRid);
		double cpuMs = RenderingServer.ViewportGetMeasuredRenderTimeCpu(_viewportRid);
		if (_subViewportRid.IsValid)
		{
			gpuMs += RenderingServer.ViewportGetMeasuredRenderTimeGpu(_subViewportRid);
			cpuMs += RenderingServer.ViewportGetMeasuredRenderTimeCpu(_subViewportRid);
		}
		string gpuStr =
			gpuMs > 0.001
				? $"GPU {gpuMs:0.00}ms -> ~{(int)(1000.0 / gpuMs)} FPS reel"
				: "GPU n/a (Metal, lance sur Linux)";
		label.Text = $"{Engine.GetFramesPerSecond()} FPS ecran | {gpuStr} | CPU {cpuMs:0.00}ms";
	}

	void OnQuitPressed()
	{
		GetTree().Root.PropagateNotification((int)NotificationWMCloseRequest);
		GetTree().Quit();
	}
}
